### PR TITLE
GH-841: Close executor in the PublisherConChannel

### PR DESCRIPTION
**Pull request by artembilan**
Fixes https://github.com/spring-projects/spring-amqp/issues/841

**cherry-pick to 2.0.x**

### CODE REVIEW

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/PublisherCallbackChannelImpl.java [hunk 1]
```diff
@@ -854,7 +862,7 @@ public synchronized int getPendingConfirmsCount(Listener listener) {
 	@Override
 	public synchronized int getPendingConfirmsCount() {
 		return this.pendingConfirms.values().stream()
-				.map(m -> m.size())
+				.map(Map::size)
```
Comment (panchenko) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/PublisherCallbackChannelImpl.java]: mapToInt() directly here?

**Comment by artembilan** [comment]
OK. No cherry-pikcing to `1.7.x`. Looks like there is no that internal executor yet...

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/PublisherCallbackChannelImpl.java [hunk 2]
```diff
@@ -817,7 +820,12 @@ public void close() throws IOException, TimeoutException {
 	}
 
 	private void shutdownCompleted(String cause) {
-		this.executor.execute(() -> generateNacksForPendingAcks(cause));
+		if (!this.executor.isShutdown()) {
+			this.executor.execute(() -> generateNacksForPendingAcks(cause));
+			if (!this.executorExplicitlySet) {
+					this.executor.shutdown();
```
Comment (panchenko) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/PublisherCallbackChannelImpl.java]: style: indentation looks wrong here.

**Comment by artembilan** [comment]
@garyrussell ,

Any thoughts how the test may fail:
```
org.springframework.amqp.rabbit.listener.DirectMessageListenerContainerMockTests > testRemoveQueuesWhileNotConnected FAILED
    org.mockito.exceptions.verification.TooLittleActualInvocations at DirectMessageListenerContainerMockTests.java:255
```
?

It doesn't for me locally 😢 

Thanks

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/PublisherCallbackChannelImpl.java [hunk 3]
```diff
@@ -119,7 +121,8 @@ public PublisherCallbackChannelImpl(Channel delegate) {
 	public PublisherCallbackChannelImpl(Channel delegate, @Nullable ExecutorService executor) {
 		delegate.addShutdownListener(this);
 		this.delegate = delegate;
-		this.executor = executor != null ? executor : DEFAULT_EXECUTOR;
+		this.executor = executor != null ? executor : Executors.newSingleThreadExecutor();
```
Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/PublisherCallbackChannelImpl.java]: No; we don't want a new executor for every channel; that's why it's static.

I suggest we move the static executor to the CCF (non-static there though) and always pass it into the ctor; then shut down the executor in the `CCF.destroy()`.

**Comment by garyrussell** [comment]
>Any thoughts how the test may fail:

Yes; it's a race in the tests; I am counting down latch3 in the mock for `basicQos` instead of `basicConsume`.

**Comment by garyrussell** [comment]
Needs a backport - many conflicts.

**Comment by artembilan** [comment]
Well, the workaround is very easy, as we all know, so let’s don’t back-port!

**Comment by garyrussell** [comment]
👍

**Pull request merged**